File: phases/15-systems-performance/13-lock-contention-patterns-and-cures/code/main.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <atomic>
#include <mutex>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <algorithm>

class SpinLock {
    std::atomic<bool> flag_{false};
public:
    void lock() {
        while (flag_.exchange(true, std::memory_order_acquire)) {
#if defined(__x86_64__) || defined(_M_X64)
            __builtin_ia32_pause();
#endif
        }
    }
    void unlock() {
        flag_.store(false, std::memory_order_release);
    }
};

class BackoffSpinLock {
    std::atomic<bool> flag_{false};
public:
    void lock() {
        int delay = 1;
        while (flag_.exchange(true, std::memory_order_acquire)) {
            for (int i = 0; i < delay; ++i) {
#if defined(__x86_64__) || defined(_M_X64)
                __builtin_ia32_pause();
#endif
            }
            delay = std::min(delay * 2, 1024);
        }
    }
    void unlock() {
        flag_.store(false, std::memory_order_release);
    }
};

class TicketLock {
    std::atomic<uint32_t> next_{0};
    std::atomic<uint32_t> serving_{0};
public:
    uint32_t lock() {
        auto ticket = next_.fetch_add(1, std::memory_order_acquire);
        while (serving_.load(std::memory_order_acquire) != ticket) {
#if defined(__x86_64__) || defined(_M_X64)
            __builtin_ia32_pause();
#endif
        }
        return ticket;
    }
    void unlock() {
        serving_.fetch_add(1, std::memory_order_release);
    }
};

struct Result {
    double elapsed_ms;
    uint64_t final_value;
    uint64_t expected;
};

template<typename LockType>
Result run_locked_benchmark(int num_threads, uint64_t per_thread, LockType& lock, std::atomic<uint64_t>& counter) {
    counter.store(0, std::memory_order_relaxed);
    auto start = std::chrono::high_resolution_clock::now();
    std::vector<std::thread> threads;
    for (int t = 0; t < num_threads; ++t) {
        threads.emplace_back([&lock, &counter, per_thread]() {
            for (uint64_t i = 0; i < per_thread; ++i) {
                std::lock_guard<LockType> lg(lock);
                counter.fetch_add(1, std::memory_order_relaxed);
            }
        });
    }
    for (auto& th : threads) th.join();
    auto end = std::chrono::high_resolution_clock::now();
    double ms = std::chrono::duration<double, std::milli>(end - start).count();
    return {ms, counter.load(), static_cast<uint64_t>(num_threads) * per_thread};
}

Result run_fetch_add_benchmark(int num_threads, uint64_t per_thread) {
    std::atomic<uint64_t> counter{0};
    auto start = std::chrono::high_resolution_clock::now();
    std::vector<std::thread> threads;
    for (int t = 0; t < num_threads; ++t) {
        threads.emplace_back([&counter, per_thread]() {
            for (uint64_t i = 0; i < per_thread; ++i) {
                counter.fetch_add(1, std::memory_order_relaxed);
            }
        });
    }
    for (auto& th : threads) th.join();
    auto end = std::chrono::high_resolution_clock::now();
    double ms = std::chrono::duration<double, std::milli>(end - start).count();
    return {ms, counter.load(), static_cast<uint64_t>(num_threads) * per_thread};
}

void print_header() {
    std::printf("\n%-15s %-8s %-12s %-12s %-8s\n", "Lock Type", "Threads", "Time (ms)", "Mops/s", "Correct");
    std::printf("%-15s %-8s %-12s %-12s %-8s\n", "--------", "------", "---------", "----------", "-------");
}

void print_result(const char* name, int threads, const Result& r) {
    double mops = r.expected / (r.elapsed_ms / 1000.0) / 1e6;
    const char* ok = (r.final_value == r.expected) ? "YES" : "NO";
    std::printf("%-15s %-8d %-12.2f %-12.2f %-8s\n", name, threads, r.elapsed_ms, mops, ok);
}

int main() {
    constexpr uint64_t INCREMENTS = 10'000'000;
    constexpr int thread_counts[] = {1, 2, 4, 8};

    std::printf("=== Lock Contention Benchmark ===\n");
    std::printf("Incrementing a shared counter %llu times per thread\n", (unsigned long long)INCREMENTS);
    std::printf("Lock types: std::mutex, SpinLock, BackoffSpinLock, TicketLock, fetch_add\n");

    print_header();

    for (int nt : thread_counts) {
        {
            std::mutex mtx;
            std::atomic<uint64_t> counter{0};
            auto r = run_locked_benchmark(nt, INCREMENTS, mtx, counter);
            print_result("std::mutex", nt, r);
        }
        {
            SpinLock spin;
            std::atomic<uint64_t> counter{0};
            auto r = run_locked_benchmark(nt, INCREMENTS, spin, counter);
            print_result("SpinLock", nt, r);
        }
        {
            BackoffSpinLock boff;
            std::atomic<uint64_t> counter{0};
            auto r = run_locked_benchmark(nt, INCREMENTS, boff, counter);
            print_result("BackoffSpin", nt, r);
        }
        {
            TicketLock ticket;
            std::atomic<uint64_t> counter{0};
            auto r = run_locked_benchmark(nt, INCREMENTS, ticket, counter);
            print_result("TicketLock", nt, r);
        }
        {
            auto r = run_fetch_add_benchmark(nt, INCREMENTS);
            print_result("fetch_add", nt, r);
        }
        std::printf("\n");
    }

    std::printf("=== Scaling Analysis ===\n");
    std::printf("Thread counts: ");
    for (int nt : thread_counts) std::printf("%d ", nt);
    std::printf("\n\n");
    std::printf("Expected behaviors:\n");
    std::printf("  1. fetch_add: best scaling — single XADD instruction, minimal contention.\n");
    std::printf("  2. SpinLock: worst scaling — CAS loop bounces cache line N times per acquire.\n");
    std::printf("  3. BackoffSpinLock: better than SpinLock — exponential backoff reduces bouncing.\n");
    std::printf("  4. std::mutex: good scaling — kernel sleep under contention, adaptive spin in glibc.\n");
    std::printf("  5. TicketLock: FIFO fairness — still spins, but no starvation.\n");
    std::printf("  6. Correctness: all approaches produce the correct final count.\n");
    std::printf("\nAs thread count increases:\n");
    std::printf("  - Lock-based throughput degrades (more time waiting, less time working).\n");
    std::printf("  - SpinLock throughput collapses (N cores fighting over one cache line).\n");
    std::printf("  - Mutex throughput plateaus (sleeping threads free up cores for useful work).\n");

    return 0;
}

File: phases/15-systems-performance/13-lock-contention-patterns-and-cures/code/main.rs
use std::sync::atomic::{AtomicBool, AtomicU32, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Instant;

struct SpinLock {
    flag: AtomicBool,
}

impl SpinLock {
    fn new() -> Self {
        SpinLock {
            flag: AtomicBool::new(false),
        }
    }
    fn lock(&self) {
        while self
            .flag
            .compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
            .is_err()
        {
            while self.flag.load(Ordering::Relaxed) {
                std::hint::spin_loop();
            }
        }
    }
    fn unlock(&self) {
        self.flag.store(false, Ordering::Release);
    }
}

struct BackoffSpinLock {
    flag: AtomicBool,
}

impl BackoffSpinLock {
    fn new() -> Self {
        BackoffSpinLock {
            flag: AtomicBool::new(false),
        }
    }
    fn lock(&self) {
        let mut delay: u32 = 1;
        while self
            .flag
            .compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
            .is_err()
        {
            for _ in 0..delay {
                std::hint::spin_loop();
            }
            delay = (delay * 2).min(1024);
        }
    }
    fn unlock(&self) {
        self.flag.store(false, Ordering::Release);
    }
}

struct TicketLock {
    next: AtomicU32,
    serving: AtomicU32,
}

impl TicketLock {
    fn new() -> Self {
        TicketLock {
            next: AtomicU32::new(0),
            serving: AtomicU32::new(0),
        }
    }
    fn lock(&self) -> u32 {
        let ticket = self.next.fetch_add(1, Ordering::Acquire);
        while self.serving.load(Ordering::Acquire) != ticket {
            std::hint::spin_loop();
        }
        ticket
    }
    fn unlock(&self) {
        self.serving.fetch_add(1, Ordering::Release);
    }
}

struct BenchResult {
    elapsed_ms: f64,
    final_value: u64,
    expected: u64,
}

fn run_mutex_benchmark(num_threads: usize, per_thread: u64) -> BenchResult {
    let counter = Arc::new((Mutex::new(0u64), AtomicU64::new(0)));
    let start = Instant::now();
    let mut handles = Vec::new();
    for _ in 0..num_threads {
        let counter = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..per_thread {
                let _guard = counter.0.lock().unwrap();
                counter.1.fetch_add(1, Ordering::Relaxed);
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    let final_val = counter.1.load(Ordering::Relaxed);
    BenchResult {
        elapsed_ms: elapsed,
        final_value: final_val,
        expected: (num_threads as u64) * per_thread,
    }
}

fn run_spinlock_benchmark(num_threads: usize, per_thread: u64) -> BenchResult {
    let lock = Arc::new(SpinLock::new());
    let counter = Arc::new(AtomicU64::new(0));
    let start = Instant::now();
    let mut handles = Vec::new();
    for _ in 0..num_threads {
        let lock = Arc::clone(&lock);
        let counter = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..per_thread {
                lock.lock();
                counter.fetch_add(1, Ordering::Relaxed);
                lock.unlock();
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    let final_val = counter.load(Ordering::Relaxed);
    BenchResult {
        elapsed_ms: elapsed,
        final_value: final_val,
        expected: (num_threads as u64) * per_thread,
    }
}

fn run_backoff_benchmark(num_threads: usize, per_thread: u64) -> BenchResult {
    let lock = Arc::new(BackoffSpinLock::new());
    let counter = Arc::new(AtomicU64::new(0));
    let start = Instant::now();
    let mut handles = Vec::new();
    for _ in 0..num_threads {
        let lock = Arc::clone(&lock);
        let counter = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..per_thread {
                lock.lock();
                counter.fetch_add(1, Ordering::Relaxed);
                lock.unlock();
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    let final_val = counter.load(Ordering::Relaxed);
    BenchResult {
        elapsed_ms: elapsed,
        final_value: final_val,
        expected: (num_threads as u64) * per_thread,
    }
}

fn run_ticket_benchmark(num_threads: usize, per_thread: u64) -> BenchResult {
    let lock = Arc::new(TicketLock::new());
    let counter = Arc::new(AtomicU64::new(0));
    let start = Instant::now();
    let mut handles = Vec::new();
    for _ in 0..num_threads {
        let lock = Arc::clone(&lock);
        let counter = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..per_thread {
                lock.lock();
                counter.fetch_add(1, Ordering::Relaxed);
                lock.unlock();
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    let final_val = counter.load(Ordering::Relaxed);
    BenchResult {
        elapsed_ms: elapsed,
        final_value: final_val,
        expected: (num_threads as u64) * per_thread,
    }
}

fn run_fetch_add_benchmark(num_threads: usize, per_thread: u64) -> BenchResult {
    let counter = Arc::new(AtomicU64::new(0));
    let start = Instant::now();
    let mut handles = Vec::new();
    for _ in 0..num_threads {
        let counter = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..per_thread {
                counter.fetch_add(1, Ordering::Relaxed);
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    let final_val = counter.load(Ordering::Relaxed);
    BenchResult {
        elapsed_ms: elapsed,
        final_value: final_val,
        expected: (num_threads as u64) * per_thread,
    }
}

fn print_header() {
    println!(
        "\n{:15} {:8} {:12} {:12} {:8}",
        "Lock Type", "Threads", "Time (ms)", "Mops/s", "Correct"
    );
    println!(
        "{:15} {:8} {:12} {:12} {:8}",
        "--------", "------", "---------", "----------", "-------"
    );
}

fn print_result(name: &str, threads: usize, r: &BenchResult) {
    let mops = r.expected as f64 / (r.elapsed_ms / 1000.0) / 1e6;
    let ok = if r.final_value == r.expected {
        "YES"
    } else {
        "NO"
    };
    println!(
        "{:15} {:8} {:12.2} {:12.2} {:8}",
        name, threads, r.elapsed_ms, mops, ok
    );
}

fn main() {
    const INCREMENTS: u64 = 10_000_000;
    let thread_counts: Vec<usize> = vec![1, 2, 4, 8];

    println!("=== Lock Contention Benchmark (Rust) ===");
    println!(
        "Incrementing a shared counter {} times per thread",
        INCREMENTS
    );
    println!("Lock types: Mutex, SpinLock, BackoffSpin, TicketLock, AtomicU64::fetch_add");

    print_header();

    for &nt in &thread_counts {
        let r = run_mutex_benchmark(nt, INCREMENTS);
        print_result("Mutex", nt, &r);

        let r = run_spinlock_benchmark(nt, INCREMENTS);
        print_result("SpinLock", nt, &r);

        let r = run_backoff_benchmark(nt, INCREMENTS);
        print_result("BackoffSpin", nt, &r);

        let r = run_ticket_benchmark(nt, INCREMENTS);
        print_result("TicketLock", nt, &r);

        let r = run_fetch_add_benchmark(nt, INCREMENTS);
        print_result("fetch_add", nt, &r);

        println!();
    }

    println!("=== Contention Scaling Summary ===");
    println!("Thread counts: {:?}", thread_counts);
    println!();
    println!("Key observations:");
    println!("  1. fetch_add (AtomicU64) is fastest — single instruction, no lock acquire/release.");
    println!("  2. SpinLock degrades severely under contention — cache-line bounce per failed CAS.");
    println!("  3. BackoffSpin improves over naive SpinLock — staggered retries reduce bouncing.");
    println!("  4. Mutex scales best under high contention — parking thread frees CPU for other work.");
    println!("  5. TicketLock ensures FIFO fairness — still spins, but no starvation guarantee.");
    println!("  6. All approaches produce correct final counts (no lost increments).");
    println!();
    println!("Scaling patterns:");
    println!("  - Ideal: throughput proportional to thread count (linear scaling).");
    println!("  - Contended lock: throughput plateaus or degrades with more threads.");
    println!("  - fetch_add: near-linear scaling for simple counter (hardware coherence).");
    println!("  - SpinLock: throughput collapses — N cores fighting over one cache line.");
    println!("  - BackoffSpin: moderate improvement — backoff reduces contention peak.");
    println!("  - Mutex: best real-world scaling under high contention — kernel scheduler helps.");
}